### PR TITLE
feat: add HTML template for Code Issue panel [IDE-180]

### PR DESCRIPTION
**Pull request by cat2608**
### Description

This PR is a continuation from  [feat: add code template #473](https://github.com/snyk/snyk-ls/pull/473), where we introduced the HTML scaffold for the **Issue Code Panel**.

The goal of this change is to migrate the **IntelliJ UI** for the Issue Code Panel to the **Language Server**. With this PR, we have replicated the following elements:

- The header of the panel, displaying the **title**, **icon**, and **vulnerability type**.
- The **description** of the vulnerability type.
- **Data Flow** section.
- **External Example Fixes** section.

### Checklist

- [ ] Tests added and all succeed
- [ ] Linted
- [ ] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

🚨After having merged, please update the CLI go.mod to pull in latest language server.

### How to test

1. Capture the **HTML** from Language Server logs in `getDetailsHtml()`
ls-logs

<details>
<summary>2. Place the **body** in the following index.html: ⬇️</summary>

```html
<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <style>
    body {
      font-family: 'Arial', sans-serif;
      background-color: #2b2b2b;
      color: #a9b7c6;
    }

    .panel-header {
      display: flex;
      flex-direction: column;
      background-color: #313335;
      padding: 16px;
      border-bottom: 1px solid #3c3f41;
    }

    .severity-container {
      display: flex;
      flex-direction: row;
      align-items: center;
      margin-bottom: 8px;
    }

    .severity-icon {
      height: 24px;
      width: 24px;
      margin-right: 8px;
    }

    .severity-text {
      font-size: 14px;
      color: #6a8759;
    }

    .severity-title {
      font-size: 18px;
      color: #a9b7c6;
      padding-left: 10px;
    }

    .issue-overview,
    .data-flow-section,
    .external-example-fixes {
      padding: 16px;
    }

    .overview-text {
      line-height: 1.6;
    }

    .data-flow-section h2,
    .external-example-fixes h2 {
      font-size: 16px;
      border-bottom: 1px solid #3c3f41;
      color: #a9b7c6;
    }

    .data-flow-body {
      background-color: #2b2b2b;
    }

    .data-flow-row {
      display: flex;
      align-items: center;
      padding: 4px 0;
      font-family: 'Courier New', monospace;
    }

    .data-flow-number {
      width: 15px;
      text-align: right;
      margin-right: 8px;
    }

    .data-flow-filepath {
      width: 130px;
      padding-right: 80px;
    }

    .data-flow-text {
      margin-right: 8px;
      padding-left: 10px;
    }

    .data-flow-delimiter {
      font-weight: bold;
      color: #cc7832;
    }

    .tabs-nav {
      display: flex;
      margin-top: 16px;
    }

    .tab-item {
      padding: 8px 16px;
      cursor: pointer;
      background-color: #3c3f41;
      border: 1px solid #3c3f41;
      margin-right: 4px;
    }

    .tab-item.is-selected,
    .tab-item:hover {
      background-color: #4c4f52;
      border-color: #4c4f52;
    }

    .tab-content {
      display: none;
    }

    .tab-content.is-selected {
      display: block;
      background-color: #2b2b2b;
      padding: 16px;
      border: 1px solid #3c3f41;
    }

    .example-line {
      font-family: 'Courier New', monospace;
      padding: 2px 4px;
      border-radius: 2px;
      display: block;
    }

    .example-line.added {
      background-color: #264d26;
      color: #ffffff;
    }

    .example-line.removed {
      background-color: #4d2626;
      color: #ffffff;
    }
  </style>
</head>

<!-- Place the generated HTML here :) -->

</html>

<script>
  const tabLinks = document.querySelectorAll('.tab-item');
  const tabContents = document.querySelectorAll('.tab-content');

  document.querySelector('.tabs-nav').addEventListener('click', (event) => {
    const clickedTab = event.target.closest('.tab-item');
    if (!clickedTab) return;

    tabLinks.forEach(tab => {
      tab.classList.toggle('is-selected', tab === clickedTab);
    });

    const targetContentId = clickedTab.id.replace('tab-link', 'tab-content');
    tabContents.forEach(content => {
      content.classList.toggle('is-selected', content.id === targetContentId);
    });
  });
</script>
```
</details>

3. Open your `index.html` in any browser:

index.thml

**Note**: 💅 The stylesheet provided is for reference purposes only. This task will be addressed in a future PR.

[IDE-180]: https://snyksec.atlassian.net/browse/IDE-180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by cmars — CHANGES_REQUESTED** [review]
[html/template](https://pkg.go.dev/html/template) should be used to generate safe HTML in Go.

> Package template (html/template) implements data-driven templates for generating HTML output safe against code injection. It provides the same interface as [text/template](https://pkg.go.dev/text/template) and should be used instead of [text/template](https://pkg.go.dev/text/template) whenever the output is HTML.

Aside from security, using templates will make the markup more readable and maintainable.

You'll be able to move a lot of the iteration and conditional logic into them as well.

File: infrastructure/code/code_html.go [hunk 1]
```diff
@@ -43,11 +45,50 @@ func getDataFlowHtml(issue snyk.CodeIssueData) string {
...
 }
 
+func getCodeDiffHtml(fix snyk.ExampleCommitFix) string {
+	linesHtml := ""
+	for _, commit := range fix.Lines {
+		linesHtml += fmt.Sprintf(`
+		<div class="example-line %s"><code>%s</code></div>`, commit.LineChange, commit.Line)
+	}
+	return linesHtml
+}
```
Comment (cmars) [inline infrastructure/code/code_html.go:60]: Example of how to use html/template. Though instead of doing this, you'll likely want to declare multiple named templates (using `{{ define ...`) and then use [ExecuteTemplate](https://pkg.go.dev/html/template#Template.ExecuteTemplate) to invoke them by name. This would be preferable to inserting rendered template strings into other templates (since they will be default HTML-escaped by design). 

```suggestion
var codeDiffTmpl = template.Must(template.New("codediff").Parse(`
{{- range .Lines }}
  <div class="example-line {{.LineChange}}><code>{{.Line}}</code></div>
{{- end }}
`)

func getCodeDiffHtml(fix snyk.ExampleCommitFix) (string, error) {
  var out bytes.Buffer
  err := codeDiffTmpl.Execute(&out, fix)
  if err != nil {
    return "", err
  }
  return out.String(), nil
}
```

**Comment by teodora-sandu** [comment]
Tested it with https://github.com/snyk/snyk-intellij-plugin/pull/496 and it looks like I'd expect it to! 
Screenshot 2024-04-02 at 12 03 08

**Comment by bastiandoetsch** [comment]
Something I noted: looks like the dataflow doesn't have links in the left part. Can we add these?

File: infrastructure/code/code_html.go [hunk 2]
```diff
@@ -68,11 +108,91 @@ func getDetailsHtml(issue snyk.Issue) string {
...
+
+	// External example fixes
+	html = replaceVariableInHtml(html, "repoCount", fmt.Sprintf("%d", additionalData.RepoDatasetSize))
+	html = replaceVariableInHtml(html, "exampleCount", fmt.Sprintf("%d", len(additionalData.ExampleCommitFixes)))
+	html = replaceVariableInHtml(html, "tabsNav", getTabsHtml(additionalData.ExampleCommitFixes))
 
 	return html
 }
+
+func getFileName(filePath string) string {
```
Comment (teodora-sandu) [inline infrastructure/code/code_html.go:131]: This function is not used anywhere now other than the tests, do we still need it?

File: infrastructure/code/code_html.go [hunk 3]
```diff
@@ -68,11 +108,91 @@ func getDetailsHtml(issue snyk.Issue) string {
...
+
+	filename := filepath.Base(filePath)
+	return filename
+}
+
+func getIssueType(additionalData snyk.CodeIssueData) string {
+	if additionalData.IsSecurityType {
+		return "Vulnerability"
+	}
+	return "Quality Issue"
```
Comment (teodora-sandu) [inline infrastructure/code/code_html.go:152]: I don't think we need to do this for `Quality Issue` types, but I could be wrong? The aim was to replace the Suggestion Panel for vulnerabilities that come up under `Code Security` in this screenshot. 
Screenshot 2024-04-02 at 13 11 41

File: infrastructure/code/code_html.go [hunk 4]
```diff
@@ -68,11 +108,91 @@ func getDetailsHtml(issue snyk.Issue) string {
...
+
+	if len(tabTitle) > 50 {
+		tabTitle = tabTitle[:50] + "..."
+	}
+
+	return tabTitle
+}
+
+func getSeverityIconSvg(issue snyk.Issue) string {
+	switch issue.Severity {
```
Comment (teodora-sandu) [inline infrastructure/code/code_html.go:174]: I don't think this is an issue but maybe Design would know best? It renders in IntelliJ so technically I don't see why we wouldn't do this

File: infrastructure/code/code_html.go [hunk 5]
```diff
@@ -59,6 +60,45 @@ func getDataFlowHtml(issue snyk.CodeIssueData) string {
...
+	}
+	return linesHtml
+}
+
+func getTabsHtml(fixes []snyk.ExampleCommitFix) string {
+	tabsHtml := `<div class="tabs-nav">`
+
+	for i, fix := range fixes {
+		// Add the is-selected class to the first tab item only
+		// The IDE handles the tab switching with the is-selected class
```
Comment (teodora-sandu) [inline infrastructure/code/code_html.go:77]: I guess the JavaScript for changing tabs was left for IntelliJ, to implement like we do for the Dataflow links in https://github.com/snyk/snyk-intellij-plugin/pull/497. 

I think that should be possible but since the interactions are all within the same panel and more "visual" based I wonder if it'd be better to include the JavaScript in the HTML. I could be wrong so feel free to challenge me! Also, if we end up deciding to put the interaction in the HTML I would leave that for a follow-up PR

File: infrastructure/code/code_html_test.go [hunk 6]
```diff
@@ -25,50 +27,261 @@ import (
...
+			Message:            "Either rethrow this java.lang.InterruptedException or set the interrupted flag on the current thread with 'Thread.currentThread().interrupt()'. Otherwise the information that the current thread was interrupted will be lost.",
 		},
 	}
 
 	// invoke method under test
-	issueDetailsPanelHtml := getDetailsHtml(issue)
+	codePanelHtml := getDetailsHtml(issue)
+
+	// assert Data Flow section
+	expectedDataFlowHeading := fmt.Sprintf("<h2>Data Flow - %d steps</h2>", len(dataFlow)) // TODO: handle pluralization
```
Comment (teodora-sandu) [inline infrastructure/code/code_html_test.go:53]: There's a TODO left here - was this addressed or still need to do? Or an improvement on the existing code maybe so we can leave it for later

File: infrastructure/code/code_html.go (hunk 2 above)
Comment (cat2608) [reply infrastructure/code/code_html.go:131]: I saw you addressed the file name with `filepath.Base` 🙌 Thank you!

I just ran `filepath.Base()` against the test cases and a few don't pass:


| Test Case                          | Input                             | Expected Output | Actual Output                    |
|------------------------------------|-----------------------------------|-----------------|----------------------------------|
| Windows_path                       | `C:\Users\johnDoe\project\file.txt` | `file.txt`     | `C:\Users\johnDoe\project\file.txt` |
| Trailing_slash                     | `/Users/johnDoe/project/`         | (empty)      | `project`                        |
| Empty_string                       | (empty)                           | (empty)      | `.`                              |
| Only_slashes                       | `/////`                           | (empty)      | `/`                              |
| Nested_directories_with_no_file    | `/a/b/c/d/e/`                     | (empty)      | `e`                              |
| Unix_root_directory                | `/`                               | (empty)      | `/`                              |

But I guess these are edge cases that we don't need to take into consideration, right?

I also tested `filepath.Base()` with `juice-shop` and it seems is working fine:

```html
<span 
  class="data-flow-filepath" 
  file-path="/Users/cata/git/juice-shop/test/api/quantityApiSpec.ts"
  start-line="227"
  end-line="227"
  start-character="8"
  end-character="17">quantityApiSpec.ts:228</span>
<span 
  class="data-flow-filepath" 
  file-path="/Users/cata/git/juice-shop/test/api/passwordApiSpec.ts"
  start-line="28"
  end-line="28"
  start-character="12"
  end-character="21">passwordApiSpec.ts:29</span>
<span 
  class="data-flow-filepath" 
  file-path="/Users/cata/git/juice-shop/test/api/deliveryApiSpec.ts"
  start-line="22"
  end-line="22"
  start-character="10"
  end-character="19">deliveryApiSpec.ts:23</span>

```

I'm deleting the function and its test 🧹

Comment (teodora-sandu) [reply infrastructure/code/code_html.go]: Hmm these are good things to consider. I wonder how this behaves now for those edge cases. I wouldn't worry for now but I'll raise a ticket to worry about later. Thanks for calling these out!

Comment (teodora-sandu) [reply infrastructure/code/code_html.go]: https://snyksec.atlassian.net/browse/IDE-231

File: infrastructure/code/code_html.go (hunk 3 above)
Comment (teodora-sandu) [reply infrastructure/code/code_html.go:135]: Hmm in that case leave this code as is. I think I need to understand this a bit better, I never even looked at the `Code Quality` section

File: infrastructure/code/code_html_test.go (hunk 6 above)
Comment (cat2608) [reply infrastructure/code/code_html_test.go:53]: I need to check how we currently handle pluralization in IntelliJ logic and I'll apply it here, thanks!!

**Comment by cat2608** [comment]
> Something I noted: looks like the dataflow doesn't have links in the left part. Can we add these?

@bastiandoetsch Could you please expand on this? Is it possible that your question is solved with the work that @teodora-sandu has done here: https://github.com/snyk/snyk-ls/pull/475 ?

File: infrastructure/code/code_html.go (hunk 2 above)
Comment (bastiandoetsch) [reply infrastructure/code/code_html.go]: I think we'll need it once the dataflow navigation is added via javascript

File: infrastructure/code/code_html.go (hunk 3 above)
Comment (bastiandoetsch) [reply infrastructure/code/code_html.go:135]: We should do both, I think :). We handle both as "Code"

File: infrastructure/code/code_html_test.go [hunk 7]
```diff
@@ -25,50 +27,261 @@ import (
...
+			name:     "GitHub URL with parameters",
+			url:      "https://github.com/juice-shop/juice-shop/commit/0fa9d5547c5300cf8162b8f31a40aea6847a5c32?diff=split#diff-7e23eb1aa3b7b4d5db89bfd2860277e5L75",
+			expected: "juice-shop/juice-shop",
+		},
+		{
+			name:     "GitHub URL with a dot in the repo name",
+			url:      "https://github.com/juice-shop/.github/commit/67603b2f2b4f02fbc65f53bda7c3f56a5d341987",
+			expected: "juice-shop/.github",
+		},
+		// TODO: Do we support non-GitHub URLs?
```
[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply infrastructure/code/code_html_test.go:135]: We don't know - that discussion is not finished, how the general representation of a target identifier would look like.

[the remark this answers is not in the document]
Comment (bastiandoetsch) [reply infrastructure/code/code_html_test.go:135]: We don't know - that discussion is not finished, how the general representation of a target identifier would look like.

File: infrastructure/code/code_html.go [hunk 8]
```diff
@@ -68,11 +108,91 @@ func getDetailsHtml(issue snyk.Issue) string {
...
 	return html
 }
+
+func getFileName(filePath string) string {
+	if filePath == "" {
+		return ""
+	}
+
+	// Windows paths
+	filePath = strings.Replace(filePath, "\\", "/", -1)
```
Comment (bastiandoetsch) [inline infrastructure/code/code_html.go]: I think, you can also use `ToSlash` for that.

File: infrastructure/code/code_html_test.go (hunk 6 above)
Comment (cat2608) [reply infrastructure/code/code_html_test.go]: Done 🙌

File: infrastructure/code/code_html.go (hunk 8 above)
Comment (cat2608) [reply infrastructure/code/code_html.go]: Adding this comment to the to IDE-231, thank you 🙏

**Review by bastiandoetsch — APPROVED** [review]
